File: getMqtt.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"sync"
	"time"

	MQTT "github.com/eclipse/paho.mqtt.golang"
)

type MetricData struct {
	Imsi string `json:"imsi"`
}

var (
	imsiCount = make(map[string]int)
	lock      sync.Mutex
)

func onConnect(client MQTT.Client) {
	fmt.Println("已連線")
	client.Subscribe("FiveGC/metric", 0, nil)
}

func onMessage(client MQTT.Client, msg MQTT.Message) {
	var data MetricData
	err := json.Unmarshal(msg.Payload(), &data)
	if err != nil {
		log.Printf("解析失敗: %v", err)
		return
	}

	if data.Imsi != "" {
		lock.Lock()
		imsiCount[data.Imsi]++
		lock.Unlock()
	}
}

func printAndReset() {
	for {
		time.Sleep(15 * time.Second)
		lock.Lock()
		if len(imsiCount) > 0 {
			fmt.Printf("這15秒有%d個不同的imsi\n", len(imsiCount))
			imsiCount = make(map[string]int)
		} else {
			fmt.Println("這15秒沒有收到訊息")
		}
		lock.Unlock()
	}
}

func main() {
	opts := MQTT.NewClientOptions()
	opts.AddBroker("tcp://10.1.153.161:1883")
	opts.SetDefaultPublishHandler(onMessage)
	opts.OnConnect = onConnect

	client := MQTT.NewClient(opts)
	if token := client.Connect(); token.Wait() && token.Error() != nil {
		log.Fatal(token.Error())
	}

	go printAndReset()

	// 保持程式運行
	select {}
}

File: getMqtt.py
import paho.mqtt.client as mqtt
import threading
import time
import json
from collections import defaultdict

imsi_count = defaultdict(int)
lock = threading.Lock()

def on_connect(client, userdata, flags, rc):
    print("已連線")
    client.subscribe("FiveGC/metric")  # 監聽你要的 topic

def on_message(client, userdata, msg):
    try:
        data = json.loads(msg.payload.decode())
        imsi = data.get("imsi")
        if imsi:
            with lock:
                imsi_count[imsi] += 1
    except Exception as e:
        print("解析失敗", e)

def print_and_reset():
    while True:
        time.sleep(15)
        with lock:
            if imsi_count:
                imsi_list = ', '.join(imsi_count.keys())
                # print(f"這5秒有{len(imsi_count)}個不同的imsi: {imsi_list}")
                print(f"這15秒有{len(imsi_count)}個不同的imsi")
                imsi_count.clear()
            else:
                print("這15秒沒有收到訊息")

client = mqtt.Client()
client.on_connect = on_connect
client.on_message = on_message

client.connect("10.1.153.167", 1883, 60)

threading.Thread(target=print_and_reset, daemon=True).start()

client.loop_forever()
